### PR TITLE
*: split out {controller,compute,storage}-types crates

### PR DESCRIPTION
**Pull request by danhhz**
This allows changes to implementation code to invalidate fewer crates, hopefully improving dev iteration times.

Notably, mz-sql and mz-transform (both heavyweight) no longer depend on storage-client. As mentioned in the reviewer notes, we could easily add mz-compute to that set.

Real-world timings:

`touch src/storage-client/src/lib.rs; cargo check`
- Before: 9.25s
- After: 8.08s

`touch src/storage-client/src/lib.rs; cargo build -p mz-environmentd -p mz-clusterd --lib`
- Before: 45.29s
- After: 22.41s
- This is synthetic for two reasons. First, I added an empty file at src/clusterd/src/lib.rs to make --lib work. Second, as the next timings show, link times still dominate.

`touch src/storage-client/src/lib.rs; cargo build --bin environmentd --bin clusterd`
- Before: 1m40s
- After: 1m39s

Deps above mz-storage-client (before)

deps-before

Deps above mz-storage-client (after)

deps-after

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Opening as a draft because there's some things to work out. Also, I got a touch sloppy with the commit hygiene and import blocks, I'll go back and tighten that up if we decide that this is desirable and if/when the final names are decided.

I feel pretty good about the foo-types model:
- There is already a mz-persist-types crate with exactly this sort of thing.
- Both mz-storage-client and mz-compute-client had types modules, which is mostly what got moved in.
- At a previous job with much worse compile times (scala), we ended up very successfully using a `foo-types`, `foo-service` (interface), `foo-concrete` model. `foo-service` corresponds to our `mz-foo-client` and `foo-concrete` corresponds to our `mz-foo`.

Names that are more or less a strawman at this point:
- I split out an mz-storage-operators to hold persist_source because it was the only reason mz-compute was depending on mz-storage-client (and we don't want to move it into mz-storage because mz-compute does not depend on mz-storage).
- As written this PR currently moves the storage controller trait out from mz-storage-client into mz-storage-types, but I think that's wrong. Instead, we should move the storage controller trait _impl_ into a new crate, something like mz-storage-controller.

Intuition for the various names and what should live in them:
- mz-storage-types, mz-compute-types: POD (plain old data), protos, no business logic (in reality, minimal business logic)
- mz-storage-client, mz-compute-client: The _adaptor_ APIs to storage and compute (aka the controllers)
- mz-storage-operators: The _compute_ API to storage
- mz-storage, mz-compute: Storage and compute dataflows

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
@benesch See "Tips for reviewer" for a bunch of context/questions. If we decide we like the idea and once we hash out names, I'd also love to hear your thoughts on which stakeholders I should pull in from the individual teams.

**Comment by benesch** [comment]
I really like the look of the "after" diagram. Unfortunately I'm not going to have time to look at the specifics of this myself! Will follow up on Slack about alternative reviewers.

**Comment by danhhz** [comment]
@antiguru @teskje @petrosagg Added y'all as individual reviewers on behalf of compute and storage. Let me know if someone else would be more appropriate. This is marked as a draft since I have some open questions (see "Tips for reviewer"), but the concept is ready for review.

Happy to answer questions, provide context, etc! I'd love to get signoffs first such that I only have to do the rebase once to land it. :)

File: src/adapter/src/coord/sequencer/linked_cluster.rs [hunk 1]
```diff
@@ -12,10 +12,8 @@
 use std::time::Duration;
 
 use mz_compute_client::controller::ComputeReplicaConfig;
-use mz_controller::clusters::{
-    ClusterId, ReplicaAllocation, ReplicaConfig, ReplicaLogging,
-    DEFAULT_REPLICA_LOGGING_INTERVAL_MICROS,
-};
+use mz_controller::clusters::{ReplicaAllocation, ReplicaConfig, ReplicaLogging};
+use mz_controller_types::{ClusterId, DEFAULT_REPLICA_LOGGING_INTERVAL_MICROS};
```
Comment (antiguru) [inline src/adapter/src/coord/sequencer/linked_cluster.rs:16]: `DEFAULT_REPLICA_LOGGING_INTERVAL_MICROS` TODO

**Review by antiguru — APPROVED** [review]
Looks great! Looked through all files, nothing stood out.

File: src/adapter/src/coord/sequencer/linked_cluster.rs (hunk 1 above)
Comment (danhhz) [reply src/adapter/src/coord/sequencer/linked_cluster.rs:16]: This const is itself a `pub use` of `mz_compute_client::controller::DEFAULT_COMPUTE_REPLICA_LOGGING_INTERVAL_MICROS` (moved to mz_compute_types in this PR), which seemed intentional, so I left it alone.

Happy to e.g. remove the re-export and inline it if you think that's better?

**Review by teskje — APPROVED** [review]
Compute changes look reasonable to me! I had a couple thoughts, but it's fine to leave them as follow ups for the compute team.

I'm wondering if there is anything we can do to be more aware of crate interdependencies in the future. I think many of us don't spend a lot of thought about adding new dependencies if convenient, and I worry that we might slowly revert to the old spiderweb if we are not careful about it.

File: src/adapter/src/coord/sequencer/linked_cluster.rs (hunk 1 above)
Comment (teskje) [reply src/adapter/src/coord/sequencer/linked_cluster.rs:16]: It is weird that the controller defines a constant that's only used by the `adapter` and `sql` crates. Seems like it should be defined in one of those instead?

**Review by danhhz — COMMENTED** [review]
> I'm wondering if there is anything we can do to be more aware of crate interdependencies in the future. I think many of us don't spend a lot of thought about adding new dependencies if convenient, and I worry that we might slowly revert to the old spiderweb if we are not careful about it.

I'd love to add a lint that prevents certain deps, but that's definitely something I'll have to leave for future skunkworks.

**Review by petrosagg — APPROVED** [review]
The `storage-types` split looks good! On `storage-operators`, is that crate pulling its weight? Is the idea behind it that it allows iterating on persist_source without recompiling `mz-storage`, which is slow?

If yes, we can also go the other way and keep `persist_source` in `mz-storage` but split out each source to its own crate. Sources already implement a trait so that ought to make things a bit better because each per-source crate will only be recompiled when that particular source changes.

Not a strong opinion on this so if you feel strongly about `mz-storage-operators` go ahead with your idea as-is. I'll experiment with splitting sources out in some future skunkworks

**Comment by danhhz** [comment]
> On storage-operators, is that crate pulling its weight? Is the idea behind it that it allows iterating on persist_source without recompiling mz-storage, which is slow?

Actually, it's almost the opposite in a way. The idea is that it allows iterating on mz-storage-client without invalidating mz-compute.

The persist_source commit is easily the least invasive of any of these, so happy to leave it off this PR, discuss async, and do whatever we decide in a followup if you prefer?

**Comment by petrosagg** [comment]
I see. It sounds like `mz_compute` should depend on `mz_storage`, just like `mz_compute_client` depends on `mz_storage_client` but that may impact linking times. Let's leave the `mz_storage_operators` as you have it!

**Comment by danhhz** [comment]
TFTRs, everyone!!

I played locally with splitting the storage controller _impl_ out into a new crate instead of splitting the _trait_ out into mz-storage-types, and I do think it's better, but it's non-obvious enough that it's worth going through a round of review. I'm going to pull that out into its own PR so I can go ahead and get this one merged

**Pull request merged**